hawkeye-api/src/backend.rs

```diff
@@ -0,0 +1,341 @@
+            &patch,
+        )
+        .await
+}
+
+pub async fn update_watcher_deployment(
+    k8s_client: &kube::Client,
+    watcher: &Watcher,
+) -> kube::Result<Deployment> {
+    // 2. Update Deployment with replicas=0
+    log::debug!("Updating Deployment instance");
+    let deploy = templates::build_deployment(
+        watcher.id.as_ref().unwrap(),
+        watcher.source.ingest_port,
+        watcher.tags.as_ref(),
+    );
+    let patch_params = PatchParams {
+        field_manager: Some(FIELD_MGR.to_owned()),
+        ..Default::default()
+    };
+    let patch = Patch::Merge(&deploy);
+    let deployments: Api<Deployment> = Api::namespaced(k8s_client.clone(), &config::NAMESPACE);
+    deployments
+        .patch(
+            &templates::deployment_name(watcher.id.as_ref().unwrap()),
+            &patch_params,
+            &patch,
+        )
+        .await
+}
+
+pub async fn update_watcher_service(
+    k8s_client: &kube::Client,
+    watcher: &Watcher,
+) -> kube::Result<Service> {
+    log::debug!("Updating Service instance");
+    let svc = templates::build_service(
+        watcher.id.as_ref().unwrap(),
+        watcher.source.ingest_port,
+        watcher.tags.as_ref(),
+    );
+    // TODO: Handle errors
+    let patch_params = PatchParams {
+        field_manager: Some(FIELD_MGR.to_owned()),
+        ..Default::default()
+    };
+    let patch = Patch::Merge(&svc);
+    let services: Api<Service> = Api::namespaced(k8s_client.clone(), &config::NAMESPACE);
+    services
+        .patch(
+            &templates::service_name(watcher.id.as_ref().unwrap()),
+            &patch_params,
+            &patch,
+        )
+        .await
+}
+
+// Apply Watcher updates and restart Watcher if it was previously running.
+pub async fn apply_watcher_updates(
+    k8s_client: &kube::Client,
+    watcher: &Watcher,
+) -> Result<(), kube::Error> {
+    // TODO: clean this up. I ran out of time trying to get ? to work correctly with warp and need
+    // to move on.
+
+    // Update k8s resources.
+    let _ = update_watcher_configmap(k8s_client, watcher).await?;
+    let _ = update_watcher_deployment(k8s_client, watcher).await?;
+    let _ = update_watcher_service(k8s_client, watcher).await?;
+
+    // Only restart the worker if it was already started.
+    let retry_sleep = time::Duration::from_secs(1);
+    log::warn!("Watcher status: {:?}", watcher.status);
+    if watcher.status != Some(Status::Ready) {
+        let watcher_id = watcher.id.as_ref().unwrap();
+        let _ = stop_watcher(k8s_client, watcher_id).await?;
+
+        // k8s can take a few seconds to update, so retry a few times.
+        for _ in 1..8 {
+            let deployment_status = get_watcher_deployment(k8s_client, watcher_id).await?;
+            // .map_err(|_| warp::reject::custom(InternalError));
+            if deployment_status.get_watcher_status() == Status::Ready {
+                break;
+            }
+            thread::sleep(retry_sleep);
```

Comment:
Since this is in an async context, we can't use the `std::thread::sleep` function, as that will block the task. Instead, the [async version from tokio](https://docs.rs/tokio/latest/tokio/time/fn.sleep.html) needs to be used:
```suggestion
            tokio::time::sleep(retry_sleep).await;
```

Reply:
If `std::thread:sleep` function would be used, then the executor thread (which continuously switches between tasks), it would be stuck waiting on `std::thread::sleep`. By using `tokio::time:sleep` instead, this task will "yield" and allow the executor to switch tasks, and allow them to make progress. Once the `retry_sleep` duration passes, this task will be "woken" and will be allowed to continue executing.

Reply:
oh sweet, makes total sense. thanks!